lean/bitvec.lean

```diff
@@ -0,0 +1,325 @@
+Authors: Joe Hendrix, Sebastian Ullrich, Jason Dagit
+
+Basic operations on bitvectors.
+
+This is a work-in-progress, and contains additions to other theories.
+-/
+import data.vector
+import .nat.lemmas
+
+import tactic.find
+
+import .tactic -- for dec_trivial_tac
+
+-- A `bitvec n` is a subtype of natural numbers such that the value of
+-- the bitvec is strictly less than 2^n.
+@[reducible]
+def bitvec (sz:ℕ) :=
+  { x : ℕ // x < (2 ^ sz) }
+
+namespace bitvec
+open nat
+open int
+open vector
+
+section zero
+
+  -- Create a zero bitvector
+  protected
+  def zero (n : ℕ) : bitvec n :=
+    ⟨0, nat.pos_pow_of_pos n dec_trivial⟩
+
+  -- bitvectors have a zero, at every length
+  instance {n:ℕ} : has_zero (bitvec n) := ⟨bitvec.zero n⟩
+
+  @[simp]
+  lemma bitvec_zero_zero (x : bitvec 0) : x.val = 0 :=
+    begin
+      cases x,
+      { simp [nat.pow_zero, lt_one_is_zero] at x_property,
+        simp, assumption
+      }
+    end
+
+end zero
+
+section one
+
+  -- Create a bitvector with the constant one, doing so requires
+  -- the bitvector to have positive length.
+  protected
+  def one_of_pos_len (n: ℕ) (H : n > 0) : bitvec n :=
+    ⟨1, calc
+          1   < 2^1 : by dec_trivial_tac
+          ... ≤ 2^n : by apply (pow_le_pow_of_le_right (zero_lt_succ _) (succ_le_of_lt H))⟩
+
+  -- In pratice, it's more useful to allow 0 length bitvectors to have 1
+  -- as well. This leads to a special case where the 0-length bitvector
+  -- 1 is really just 0. This turns out to simplify things.
+  protected
+  def one : Π(n:ℕ), bitvec n
+  | 0        := bitvec.zero 0
+  | (succ m) := bitvec.one_of_pos_len (succ m) (nat.zero_lt_succ m)
+
+  instance {n:ℕ} : has_one (bitvec n)  := ⟨bitvec.one n⟩
+
+end one
+
+protected def cong {a b : ℕ} (h : a = b) : bitvec a → bitvec b
+| ⟨x, p⟩ := ⟨x, h ▸ p⟩
+
+-- Most significant bit
+def msb {n:ℕ} (x: bitvec n) : bool := (nat.shiftl x (n-1)) = 1
+
+-- 2s complement negation
+protected def neg {n:ℕ} (x : bitvec n) : bitvec n :=
+  ⟨if x.val = 0 then 0 else 2^n - x.val,
+   begin
+     by_cases (x.val = 0),
+     { simp [h], apply pos_pow_of_pos, dec_trivial_tac },
+     { simp [h],
+       have pos : 0 < 2^n - x.val, { apply nat.sub_pos_of_lt x.property },
+       have x_val_pos: 0 < x.val, { apply nat.pos_of_ne_zero h },
+       apply sub_lt_of_pos_le x.val (2^n) x_val_pos,
+       apply le_of_lt x.property,
+     }
+    end⟩
+
+section conversion
+  -- Operations for converting to/from bitvectors
+  variable {α : Type}
+
+  protected def to_nat {n : nat} (v : bitvec n) : nat := v.val
+
+  protected def to_int {n:ℕ} (x: bitvec n) : int :=
+    match msb x with
+    | ff := int.of_nat x.val
+    | tt := int.neg_of_nat (bitvec.neg x).val
+  end
+
+  protected def of_nat (n : ℕ) (x:ℕ) : bitvec n := ⟨ x % 2^n, by simp ⟩
+
+  protected def of_int : Π (n : ℕ), int → bitvec (succ n)
+  | n (int.of_nat m)          := bitvec.of_nat (succ n) m
+  | n (int.neg_succ_of_nat m) := bitvec.neg (bitvec.of_nat (succ n) m)
```

Comment:
The original may have had this type, but I'd think `of_int` could just have the type `\Pi (n:\Nat), int \to bitvec n`, and we define the zero case to be always 0.  Otherwise, we always have to remember to subtract one, e.g., `of_int 63 x` to get a 64-bit bitvector, and that looks weird to me.